controllers/webhook.go

```diff
@@ -0,0 +1,166 @@
+
+	csvWebhook := admrv1.MutatingWebhook{
+		Name: WebhookName,
+		ClientConfig: admrv1.WebhookClientConfig{
+			Service: &admrv1.ServiceReference{
+				Name: WebhookServiceName,
+				Path: ptr.To(odfwebhook.WebhookPath),
+				Port: ptr.To(int32(443)),
+			},
+		},
+		Rules: []admrv1.RuleWithOperations{
+			{
+				Rule: admrv1.Rule{
+					APIGroups:   []string{"operators.coreos.com"},
+					APIVersions: []string{"v1alpha1"},
+					Resources:   []string{"clusterserviceversions"},
+					Scope:       ptr.To(admrv1.NamespacedScope),
+				},
+				Operations: []admrv1.OperationType{admrv1.Create},
+			},
+		},
+		SideEffects:             ptr.To(admrv1.SideEffectClassNone),
+		TimeoutSeconds:          ptr.To(int32(30)),
+		AdmissionReviewVersions: []string{"v1"},
+		// fail the validation if webhook can't be reached
+		FailurePolicy: ptr.To(admrv1.Fail),
+	}
+
+	whConfig := &admrv1.MutatingWebhookConfiguration{}
+	whConfig.Name = csvWebhook.Name
+
+	res, err := controllerutil.CreateOrUpdate(ctx, cli, whConfig, func() error {
+
+		// openshift fills in the ca on finding this annotation
+		whConfig.Annotations = map[string]string{
+			"service.beta.openshift.io/inject-cabundle": "true",
+		}
+
+		var caBundle []byte
+		if len(whConfig.Webhooks) == 0 {
+			whConfig.Webhooks = make([]admrv1.MutatingWebhook, 1)
+		} else {
+			// do not mutate CA bundle that was injected by openshift
+			caBundle = whConfig.Webhooks[0].ClientConfig.CABundle
+		}
+
+		// webhook desired state
+		var wh *admrv1.MutatingWebhook = &whConfig.Webhooks[0]
+		csvWebhook.DeepCopyInto(wh)
+
+		wh.Name = whConfig.Name
+		// only send requests received from own namespace
+		wh.NamespaceSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"kubernetes.io/metadata.name": operatorNamespace,
+			},
+		}
+		// preserve the existing (injected) CA bundle if any
+		wh.ClientConfig.CABundle = caBundle
+		// send request to the service running in own namespace
+		wh.ClientConfig.Service.Namespace = operatorNamespace
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	logger.Info("successfully created or updated webhook", "operation", res, "name", whConfig.Name)
+	return nil
+}
+
+func reconcileCsvMutatingWebhookService(ctx context.Context, cli client.Client, logger logr.Logger, operatorNamespace string) error {
+
+	webhookService := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      WebhookServiceName,
+			Namespace: operatorNamespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "https",
+					Port:       443,
+					Protocol:   corev1.ProtocolTCP,
+					TargetPort: intstr.FromInt32(9443),
+				},
+			},
+			Selector: map[string]string{
+				"app.kubernetes.io/name": "odf-operator",
+			},
+			Type: corev1.ServiceTypeClusterIP,
+		},
+	}
+
+	svc := &corev1.Service{
+		ObjectMeta: webhookService.ObjectMeta,
+	}
```

Comment:
this is recreated in every func invocation, why can't move this to a template and deep copy whatever is required?

Reply:
added as a variable.